src/main/java/uk/gov/pay/products/resources/ProductResource.java

```diff
@@ -88,6 +105,38 @@ public Response findProductByGatewayAccountIdAndExternalId(@PathParam("gatewayAc
                         Response.status(NOT_FOUND).build());
     }
 
+    @POST
+    @Path("/products/{productExternalId}/regenerate-api-key")
+    @Produces(APPLICATION_JSON)
+    @Consumes(APPLICATION_JSON)
+    public Response regenerateProductApiKey(@PathParam("productExternalId") String productExternalId) {
+        return productFactory.productFinder().findByExternalId(productExternalId).map((product) -> {
```

Comment:
I think most of the logic here should be in a service rather than directly in the resource. The resource should really just call the service, which would do the actual work of regenerating the token, with the resource just producing the appropriate response based on whether or not it was successful.

A separate service would also be easier to unit test, which might mean we need fewer integration tests.